tfhe/src/core_crypto/entities/modulus_switched_lwe_ciphertext.rs

```diff
@@ -0,0 +1,286 @@
+///
+/// assert_eq!(
+///     modulus_switch(
+///         decrypt_lwe_ciphertext(&lwe_secret_key, &lwe_ms_ed).0,
+///         CiphertextModulusLog(5)
+///     ),
+///     0
+/// );
+/// ```
+pub struct PackedModulusSwitchedLweCiphertext<Scalar: UnsignedTorus> {
+    packed_coeffs: Vec<Scalar>,
+    lwe_dimension: LweDimension,
+    log_modulus: CiphertextModulusLog,
+    uncompressed_ciphertext_modulus: CiphertextModulus<Scalar>,
+}
+
+impl<Scalar: UnsignedTorus> PackedModulusSwitchedLweCiphertext<Scalar> {
+    /// Compresses a ciphertext by reducing its modulus
+    /// This operation adds a lot of noise
+    pub fn compress<Cont: Container<Element = Scalar>>(
+        ct: &LweCiphertext<Cont>,
+        log_modulus: CiphertextModulusLog,
+    ) -> Self {
+        let switch_modulus = |x| modulus_switch(x, log_modulus);
+
+        let log_modulus = log_modulus.0;
+
+        let uncompressed_ciphertext_modulus = ct.ciphertext_modulus();
+
+        assert!(
+            ct.ciphertext_modulus().is_power_of_two(),
+            "Modulus switch compression doe not support non power of 2 input moduli",
+        );
+
+        let uncompressed_ciphertext_modulus_log =
+            if uncompressed_ciphertext_modulus.is_native_modulus() {
+                Scalar::BITS
+            } else {
+                uncompressed_ciphertext_modulus.get_custom_modulus().ilog2() as usize
+            };
+
+        assert!(
+            log_modulus <= uncompressed_ciphertext_modulus_log,
+            "The log_modulus (={log_modulus}) for modulus switch compression must be smaller than the uncompressed ciphertext_modulus_log (={uncompressed_ciphertext_modulus_log})",
+        );
+
+        let lwe_size = ct.lwe_size().0;
+
+        let number_bits_to_pack = lwe_size * log_modulus;
+
+        let len = number_bits_to_pack.div_ceil(Scalar::BITS);
+
+        let slice = ct.as_ref();
+
+        //                          log_modulus
+        //                           |-------|
+        //
+        // slice        :    |   k   |  k+1  |  k+2  |
+        // packed_coeffs:  i-1   |       i       |     i+1
+        //
+        //                       |---------------|
+        //                         Scalar::BITS
+        //
+        //                   |---|
+        //                start_shift
+        //
+        //                       |---|
+        //                       shift1
+        //                (1st loop iteration)
+        //
+        //                       |-----------|
+        //                           shift2
+        //                    (2nd loop iteration)
+        //
+        // packed_coeffs[i] =
+        //                    slice[k] >> start_shift
+        //                    | slice[k+1] << shift1
+        //                    | slice[k+2] << shift2
+        //
+        // with << corresponding to shifting bits on the right
+        // and  >> corresponding to shifting bits on the left
```

Comment:
I think this is more confusing than it needs to

<< is a left shift
>> is a right shift

I think you can remove those two lines, or use the "classic" wording for shifts :)

Reply:
I don't think so
In the diagram, the lowest bits are on the left which means a left shift shifts bits to the right and conversely 
Hence the need for this comment

Reply:
I don't get it 🤔